rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,500 @@
+
+## Detailed design
+
+TypeScript should be treated with exactly the same rigor as other elements of the Ember ecosystem, with the same level of commitment to stability, clear and accurate use of Semantic Versioning, testing, and clear policies about breaking changes.
+
+TypeScript introduces two new concerns around breaking changes for addons which publish types.
+
+1.  TypeScript does not adhere to the same norms around Semantic Versioning as the rest of the npm ecosystem, so it is important for addon authors to understand when TypeScript versions may introduce breaking changes without any other change made to the addon.
+
+2.  The runtime behavior of the addon is no longer the only source of potentially-breaking changes: types may be as well. In a well-typed addon, runtime behavior and types should be well-aligned, but it is possible to introduce breaking changes to types without changing runtime behavior.
+
+Accordingly, we must define breaking changes precisely and carefully.
+
+### Background: TypeScript and Semantic Versioning
+
+TypeScript ***does not*** adhere to Semantic Versioning, but since it participates in the npm ecosystem, it uses the same format for its version numbers: `<major>.<minor>.<patch>`. In Semantic Versioning, `<major>` would be a breaking change release, `<minor>` would be a backwards-compatible feature-addition release, and `<patch>` would be a “bug fix” release.
+
+TypeScript maintains this final distinction: the `<patch>` position corresponds to bug fix changes released between the time of one `<minor>` release and the next. For example, `3.9.3` is the third patch/bug fix release for TypeScript version 3.9. However, both `<major>` and `<minor>` releases *may* introduce breaking changes of the sort that the npm ecosystem normally reserves for the `<major>` slot in the version number. (Not all `<minor>` *or* `<major>` releases *do* introduce breaking changes in the normal Semantic Versioning sense, but either *may*.) Accordingly, and for simplicity, the Ember ecosystem should treat *all* TypeScript `<major>.<minor>` releases as *potentially* representing breaking changes.
+
+By example:
+
+-   `3.9.2` to `3.9.3`: always considered a bug fix
+-   `3.9.3` to `4.0.0`: *may or may not* introduce breaking changes
+-   `3.8.3` to `3.9.0`: *may or may not* introduce breaking changes
+
+### Defining breaking changes
+
+When making a change to the types of the public interface, it is subject to the same constraints as runtime code: *breaking the published types entails a breaking change.* Not all changes to published types are *breaking*, however: some changes will continue to allow user code to continue working without any issue. Thus, a *breaking* change to types is a change where previously-allowed code will stop type-checking.
```

Comment:
To expand on Mike's point, 'backwards compatibility' seems like a difficult (or maybe just impossible) thing to define in a vacuum. There was a bit of [discussion on the `semver` repo](https://github.com/semver/semver/issues/256) a while back about whether compatibility applies to documented behavior, observable behavior, or something else entirely, and thread never really reached a conclusion. My loose takeaway is that norms for backwards compatibility are informed both by the technical details of the system in question and by the expectations of the surrounding community.

"All observable behavior" seems like a non-starter as the bar for compatibility, since you then wind up at a nihilistic [_every single release is a breaking change_](https://xkcd.com/1172/) perspective.

On the other hand, restricting your compatibility guarantees to "Only Documented Patterns" strikes me as tricky to implement in practice, since part of the purpose of shipping type declarations with a library is to document intended usage patterns.

As a concrete example based on [this thread below](https://github.com/typed-ember/ember-cli-typescript/pull/1158#discussion_r453283441), consider this code:

```ts
declare function go(): { a: string; b?: number };
```

If I were to make `b` required in the return type, that wouldn't break code that calls `go`—the new return type is still assignable to the old one, so in principle you're dropping useful type information on the floor and making checks that are no longer necessary, but your code should still typecheck.

What if I write something like this as a consumer of that library, though?

```ts
let value: ReturnType<typeof go> = { a: 'hello' };
```

Making `b` required in the return type of `go` would cause that code to stop typechecking.

My personal expectation as a consumer is that that would nevertheless be treated as a non-breaking change. I could easily imagine people disagreeing with that assessment, though, and others of you commenting on this PR may! This is what I mean by how community norms factor in to calculating what is and isn't breaking.

Part of the purpose of this document, of course, is to help set those norms, but it seems like there may not yet be complete agreement about what they'll be.

Reply:
Yeah, thanks for framing it so well @mike-north and @dfreeman! One of the keys I see as I'm digging into this particular question is the dynamic quality of TypeScript's types. What follows is basically my "thinking out loud" a bit in this space, so it's very possible I miss other examples or fail to capture important details in this specific writeup—please flag them if you note them.

In general, other type systems (whether nominal or structural) can specify what constitutes a breaking change much more straightforwardly, because the type systems are (for lack of a better word) much more *static*. You would never have a type like `ReturnType<T>` in those languages for two reasons:

-   In general, the return type of a given function would always be well-known and static. For the general case, the most "dynamic" the return type could get would be returning an `interface` rather than a concrete class. 

-   The type systems do not even have the relevant primitives to express something like `ReturnType<T>`. TypeScript's conditional types (as well as mapped types, its keyof operator, etc.) are cutting-edge type system features which are *necessary* to express the extremely wide range of dynamic behaviors of JavaScript.
    
That is, it is specifically the combination of these advanced type system features with TypeScript's structural typing which allows users to *construct* types in ways that would be impossible in other languages. The motivating example Mike and Dan gave is a perfect example of this (and is also suggestive of one possible answer to the question). For a user to do the same sort of thing in a language like Java or C#, they would have to explicitly implement an interface with their own class to fulfill that contract.

The only other languages which have these kinds of facilities are fully-dependently-typed languages. To my knowledge, there is little discussion of how SemVer applies in the context of languages where values can be lifted to have type-level semantics or vice versa; we're very much in uncharted waters here.

An additional wrinkle is that our own choices around type definitions to date have *encouraged* users to make heavy use of those patterns. For example, because we have reasonably chosen no to define imports for types which cannot be imported in Ember (e.g. `Transition`), users have had to either import them from private locations or use a type like `ReturnType<T>` to define an alias for the type locally:

```ts
import type RouterService from '@ember/routing/router-service';
type Transition = ReturnType<RouterService['transitionTo']>
```

`ReturnType<T>` is an interesting and illuminating case: the only scenario where I would expect a user to legitimately be constructing a return type *themselves* is in testing. This is, to be clear, a legitimate use of the types. However, it's also clarifying: if the user expects to be constructing the *same* type as is returned by the function, e.g. for mocking, and that type changes, then it is a breaking change exactly and only when we *already* define it as a breaking change in the specific definitions below when discussing what constitutes a breaking change to the type of a function:

> - a function (including a class constructor or method) *returns a less specific (“wider”) type*, for example if it previously returned `string` but now returns `string | null`—since the user’s existing handling of the return value will be wrong in some cases

This is precisely the scenario shown in Dan's example code: the function now returns a wider type. The fact that a user was getting at the type via `ReturnType<T>` rather than a named `interface` doesn't seem to me to alter the semantics of the change here.

Likewise, when considering mapped types, including `keyof`-powered types, which are another possible source of this kind of dynamicism, we can cover it straightforwardly, by capturing some of the cases @mike-north called out as missing: changing the properties on an object-like definition in breaking ways, e.g. adding a new required property, removing existing required properties, etc.

Thinking about it in these terms leads me to conclude that the key is explaining in this paragraph that breaking changes *cannot* be short-handed, but only specifically in terms of the definitions we supply. That would lead me to rewrite this paragraph to read something like this:

> When making a change to the types of the public interface, it is subject to the same constraints as runtime code: *breaking the published types entails a breaking change.* Not all changes to published types are *breaking*, however: some changes will continue to allow user code to continue working without any issue. However, it is impossible to define the difference between breaking and non-breaking changes purely in the abstract.
> 
> Instead, we must define exactly what changes are “backwards-compatible” and which are “breaking,” and we must further define what constitutes a legitimate “bug fix” for type definitions designed to represent runtime behavior. We propose the following specific definitions of breaking, non-breaking, and bug-fix changes for types in the Ember community:

Reply:
Update: I still think that revised paragraph (which I've pushed) captures the sentiment correctly, but I reread the example with `ReturnType` and have reversed my view on that specific example. We may need to put in special casing for some of the utility types—I think mapped types would have a similar issue in terms of the narrowing/widening behaviors as outlined in the detailed design below. I'll keep chewing on that particular problem, while elaborating what we do have to capture the other cases Mike raised. Proposals extremely welcome.

(I note that @dfreeman's prediction about people disagreeing on this has now been fulfilled by a single person disagreeing *with himself* in the span of less than an hour. 😂)

Reply:
And now I'm going back on my going back. I think the construction-of-a-return-type scenario is *odd*, and again the only place I think it's reasonable for a user to be doing something like this is in test code; and I think if you're *constructing* a return type to match the producer (vs. an explicit interface), and specifically in the case where the only public API for constructing a type is a function provided by the framework, it's fair to say that it's *not breaking* for the framework to make a change that will never break runtime code. These types are by definition *not*  meant to be constructed by users, otherwise we would supply a class which they could subclass.

This does raise a question we'll want to consider when thinking about how this will interact with API designs for addons shipping types, up to and including core Ember addons: public `interface` types will need to specify whether they're designed to be user-constructible or not. It's possible to constrain that to some degree using brands to create nominal-ish interfaces, but that comes with its own downsides for testing code, and may require unsafe casts, which in turn creates a new source of breakage.

*«sigh»*

…I will keep thinking on this.

Reply:
A further point to consider: if the library exported the type as well as the function:

```ts
export interface Neat {
  a: string;
  b?: number;
}

declare function go(): Neat;
```

Then if we changed the exported type of `Neat` as suggested here:

```diff
export interface Neat {
  a: string;
- b?: number;
+ b: number;
}

declare function go(): Neat;
```

Then we would treat this as a breaking change for consumers of the library, precisely because they could implement it. I'm *slightly* inclined to say that *this* is precisely the fine distinction we want to make: if the type itself is exported, then making the optional property required is a breaking change; if the type is *not* exported, and is available only via `ReturnType`, then making the optional property required is *not* a breaking change. That makes a certain amount of intuitive sense. But it runs right back into the problem of allowing users to conveniently name types they're not intended to construct, and the nominal-ish interface workaround has the aforementioned problem for test code.

Reply:
I've spent further time thinking on it and talking it over with a friend who's used TS for the last four years but *outside* the Ember community, @bmakuh. His intuition matched my own:

- if an interface is exported, the change described here is breaking
- if no interface is exported, the change described here is *not* breaking

He framed this in terms of ownership and responsibility: in the case where there is no exported interface, the function very clearly has the responsibility of constructing the type, and callers are *not* expected to be able to construct an instance themselves—even if structural typing + `ReturnType<T>` means it's possible to do so. When exporting an interface, however, you're giving responsibility to others: "This is a shape you can construct."<sup>see note</sup>

From a close reading of your comments on what constitutes breaking and non-breaking changes here, as well as past discussions with @rwjblue and @pzuraq, I suspect this will closely match *most* people’s intuitions. Both @jamescdavis and @dfreeman suggested that they would *not* expect the `ReturnType<T>` change to be breaking, and I agree. At the same time, @mike-north noted, and I strongly agree (and believe Dan and James do as well) that changing the type on an exported interface<sup>see note 2</sup> in this way *is* breaking.

In sum, I think the rule should be: **narrowing a type exported as an interface or type alias is breaking; narrowing a type *only* returned from a function is *not* a breaking change.**

There are a few ways of threading the needle when it comes to types like `Transition` and `RouteInfo` (even in general libraries):

- Continue to simply advise that users create their own local aliases for these types. This is annoying, but it works.

- We can find some way of specifying that a given interface is not meant for end users to construct, and should *only* be used to name types returned from the library. Since it's defined as private API, changes of this sort are still legal, and if users *are* constructing it, they’re responsible to absorb breaking changes. The risk here is users will construct these types themselves anyway; in many cases this will end up being “intimate API.”

- For cases where a library author wants to provide a strong degree of rigor, the library author can use nominal-ish types, export the interface, and provide test helpers which solve the problem of constructing in tests.

- Go ahead and publish an interface, documenting that it's not actually meant to be user-constructed, but own the fact that as a published interface making changes to it *will* be breaking… so publish as a breaking change when you make a breaking change.

Perhaps our best option is to define the rule as given above, and to provide these options to end users?

---

**Notes:**

1. As a point of interest, @pzuraq recently noted to me that *not* enforcing this notion of ownership has led to some difficulties in maintainability in some parts of the Glimmer codebase, because TS's structural typing meant that it was possible to create a given type anywhere, and therefore people *did*.
2. I could have sworn I *had* these object-driven examples per Mike's comments; it's possible I messed up something in Git at some point. Or maybe I had just mapped them out and failed to write them down. :man_shrugging:

Reply:
> Perhaps our best option is to define the rule as given above, and to provide these options to end users?

I think this is probably the most pragmatic way to go, given the inherent tension in the different constraints here. As both an author and a library consumer, I have a pretty strong personal preference for your third bullet (best-we-can-fake-it nominality), but I suspect it's pretty likely that people will organically find themselves many of the other possible scenarios and it's important to make clear what is and isn't a breaking change once you're there.

Also potentially worth noting that `export type` makes it much easier to only export the type side of a class declaration than it used to be (when you had to do aliasing shenanigans)

Reply:
Addendum to this discussion: I also chatted with @rwjblue, @pzuraq, and @cafreeman about it, and their comments broadly matched what we'd hit on, with the additional note from Rob that "Documented as public or private" is also a good enough bar: if you're depending on an interface which is exported, but it's explicitly private, it's still legitimate to make a breaking change. This gets at Mike's original point above in terms of keeping these norms aligned with community norms.

However, I'll still note in the writeup that where possible people may want to reach for tools like API Extractor to be able to distinguish between *exported* types (i.e. available for import throughout the library) and *published* types (consumable by *users* of the library).

Reply:
All right, I've added examples covering this as well as a discussion of how users can mitigate it. I'm marking this as resolved accordingly – thanks for the detailed back and forth here! I think it captured a key distinction that I hadn't worked out previously.